### ISSUE TITLE
Errors format is too complex

### ISSUE THREAD

**Issue by Gr1N** [samuelcolvin/pydantic#162]
Hi,

I would like to discuss errors format, to my mind it's too complex to understand and parse for now.

1. Let's start from something simple
```
class Test(BaseModel):
    k_str: constr(min_length=42)
    k_int: int

Test(k_str='min')
```

1.1 Error codes only
```
{
    "k_str": "min_length",
    "k_int": "required",
}
```

1.2 Error codes + human readable messages
```
{
    "k_str": {
        "error_code": "min_length",
        "error_message": "...",
    },
    "k_int": {
        "error_code": "required",
        "error_message": "...",
    },
}
```

I prefer variant 1.1 and in all examples below I will use it.

2. Recursive models
```
class Foo(BaseModel):
    count: int

class Bar(BaseModel):
    uid: uuid.UUID

class Spam(BaseModel):
    foo: Foo
    bars: List[Bar]

Spam.parse_obj({
    'bars': [
        uuid.uuid4(),
        'not_uuid',
    ],
})
```

2.1 Error codes only
```
{
    "foo": {
        "count": "required",
    },
    "bars": {
        1: {
            "uid": "invalid",
        },
    },
}
```

Some points about format:
1. We need `error_code` to have ability easily write error parsers, it's very useful for APIs and documentation. We can extend error codes and add more context, e.g.: `min_length:42` where `42` is required minimum.
2. We need `error_message` in some cases for humans, but I don't think that you wound like to think about localization... Also I think any developer can build their own human readable messages using error codes only.
3. I don't think that we need to return list of errors for one type, let's return first one. It's more easily to implement and support and it's better for performance reasons.
4. Do not use `track` attr for recursive models, it's complex to parse. Let's use tree logic and dictionaries can help us with this. Lists can be represents as dictionaries too, just use element index as key.
5. Using my proposal you can easily convert errors dictionary to something else e.g. to JSONPoiters https://tools.ietf.org/html/rfc6901.
6. This format isn't my idea, something very similar used in [Cerberus](http://docs.python-cerberus.org/en/stable/index.html).

I know that changes like this are backward incompatible, so let's discuss and let me know if you need more detailed examples. Also I can try to implement proof of concept and send PR to you.

**Comment by samuelcolvin** [comment]
Total age errors need a cleanup, I added it to #143 buy haven't started it yet. 

I think basically errors should contain two things:

* A human readable message, suitable for end users or to help developers understand what they've done wrong
* A code, suitable for programatic use, a bit like an http status code.

We might also need a way to identify where errors happen in complex types eg. Index of lists or key of dicts.

**Comment by Gr1N** [comment]
Totally agree and my proposal covers all your points.

**Comment by Gr1N** [comment]
@samuelcolvin I spent some time to understand how library works and I think that error format implemented by you is good enough and my proposal is not (to be honest now I think it's bad).

But I found what I want to improve and same things described in #176 and #177. To solve my problems I need an ability to override default error messages.

From my point of view I see two possible solutions:
1. Add a way to override `error_msg` by type. 
We can define default error messages globally using config, something like this:
```
class BaseConfig:
    error_messages = {
        'str': {
            'min_length': 'length less than minimum allowed: {min_length}',
        },
    }
```
Also we can add ability to specify error messages by type:
```
class ConstrainedStr(str):
    error_messages = {...}
```
And find a way to pass error messages to `TypeError` and `ValueError` exceptions.

2. Add a way to override error messages globally and by type as described in option 1, but also change a bit error format by adding `error_code` which can be very useful in case of building APIs. I see only one way to do this is by introducing `FieldError` exception instead of `TypeError` and `ValueError`. 

What do you think? To my mind we can try to implement just ability to override error messages and having this ability any developer can change messages to codes.

**Comment by samuelcolvin** [comment]
For me errors should be made up of three things:

### 1 the error code

This should be a unique identifier of the error readable by a computer. 

I think the best way of building this would be something like:
* `value_error.size.below` if a value is too small
* `value_error.decimal_infinite` for a decimal which is infinite
* just `type_error` if the type is wrong and we don't know any more than that

I'm obviously flexible on the names etc., but in summary:
* all error codes start with `value_error` or `type_error` (all field errors should be `ValueError` or `TypeError` or subclasses thereof)
* this code can be inferred from the exception, but most exceptions will need to be modified to add more information
* The error can be processed by looking for particular codes or by doing one thing for `value_error.*` and something else for `type_error.*` etc.

### 2 the error's location

Made up of:
* field name, perhaps this can be "global" in some cases, eg. when validators look at multiple fields
* index (possibly) if the error occurs in a list
* the subfield (possibly) if the error occurs in a sub model or dict

This can be recursive, perhaps something like `foo_field.0.name` or just `foo_field`

### 3 the error message context

values required to render the message, eg. `{'min_length': 2}` in your example above.

-------------------------------------------------

The job of taking that data and deciding how to render the error is a different thing:
* you might want to use custom error messages.
* you might want to use gettext and translate the error messages
* you might want to remove everything except the field name and error message and return a simple dict, which your js can parse and apply to a form
* you might know your API is well documented and not bother formatting the error message at all, and just return the raw data `type`, `location`, `context`
* you might even want to inspect an input file and add line numbers to errors like [trafaret-config](https://github.com/tailhook/trafaret-config)

In short, information about the errors and how you want to render/format them should be completely separated. Of course, the `__str__` method on `ValidationError` should still show something helpful, but that's just rendering.

I guess we need a central dictionary `type > message_template` for looking up messages from types, which users can then override in their own code.

Does this make sense?

**Comment by Gr1N** [comment]
Thank you for such detailed answer. I have no questions for now and I will start from part one "the error code". Wait for PR...

**Comment by samuelcolvin** [comment]
2 more points:

Errors, as stored on a `ValidationError` should be a list:

```py
[
    {
        'type': 'value_error.size.below',
        'loc': 'foobar',
        'ctx': {'min_length': 2}
    },
    {
        'type': 'value_error.decimal_infinite',
        'loc': 'whatever.0',
    },
]
```

We need to think about how type is inferred from the exception. Maybe `<snake case root type>.<snake case class name>.<some field on the exception>`?

**Comment by samuelcolvin** [comment]
or we could stick with camel case?

**Comment by Gr1N** [comment]
> We need to think about how type is inferred from the exception. Maybe <snake case root type>.<snake case class name>.<some field on the exception>?

I think about using `cls.__mro__` to get chain of exceptions, then join them and convert to snake case, something like:
```
In [1]: class MyError(TypeError):
   ...:     pass
   ...:
   ...:

In [2]: exc = MyError('error message')

In [3]: chain = type(exc).__mro__

In [4]: chain
Out[4]: (__main__.MyError, TypeError, Exception, BaseException, object)

In [5]: chain = chain[:2]

In [6]: chain
Out[6]: (__main__.MyError, TypeError)

In [7]: chain = chain[::-1]

In [8]: chain
Out[8]: (TypeError, __main__.MyError)

In [9]: import re

In [10]: def to_snake_case(value):
    ...:     value = re.sub(r"([A-Z]+)([A-Z][a-z])", r'\1_\2', value)
    ...:     value = re.sub(r"([a-z\d])([A-Z])", r'\1_\2', value)
    ...:     value = value.replace("-", "_")
    ...:     return value.lower()
    ...:
    ...:

In [11]: to_snake_case('.'.join([t.__name__ for t in chain]))
Out[11]: 'type_error.my_error'

In [12]:```

**Comment by samuelcolvin** [comment]
Looks good, except I think you can do `to_snake_case` with just one regex: 

```py
re.sub(r'([a-z])([A-Z])', r'\1_\2', '.'.join(['ValueError', 'MyError'])).lower()
```

**Comment by Gr1N** [comment]
Things to do from #179 
* https://github.com/samuelcolvin/pydantic/pull/179#discussion_r190188299
* https://github.com/samuelcolvin/pydantic/pull/179#discussion_r190188459
* https://github.com/samuelcolvin/pydantic/pull/179#discussion_r190189190

**Comment by Gr1N** [comment]
@samuelcolvin I would like to discuss ways of errors context implementation. Let's discuss in case of `Decimal` type errors.

As I understand earlier we want to introduce specific errors, something like:
```
# current implementation
raise ValueError(f'ensure that there are no more than {cls.max_digits} digits in total')

# possible implementation
class DecimalMaxDigitsError(ValueError): pass
raise DecimalMaxDigitsError()
```

Also we want to pass error context, for example, maximum count of digits:
```
raise DecimalMaxDigitsError(max_digits=...)
```

But we can't call `DecimalMaxDigitsError` with `kwargs`, so we need to redefine `__init__`:
```
class DecimalMaxDigitsError(ValueError): 
    def __init__(self, *, max_digits):
        super().__init__(self)
        self.max_digits = max_digits
```
And I think this is ugly... Don't forget we need one more method in exception class to return context as dict somehow.

Plus where we want to store error classes? Near types/validators or in exceptions module?

Please share your thoughts if you have any.

**Comment by samuelcolvin** [comment]
Two options:

```py
class PydanticValueError(ValueError):
    def __init__(self, type, **ctx):
        self.type = type
        self.ctx = ctx
        super().__init__()

raise PydanticValueError('decimal_max_digits', max_digits=123)
```

or

```py
class PydanticValueError(ValueError):
    def __init__(self, **ctx):
        self.ctx = ctx
        super().__init__()

class DecimalMaxDigitsError(PydanticValueError): pass

raise DecimalMaxDigitsError(max_digits=123)
```

The first is simpler, the second is more pythonic. In both cases we can implement `__str__` so the errrors render properly.

What do you think?

**Comment by Gr1N** [comment]
Something like `errors.py`?

**Comment by Gr1N** [comment]
Let's try gitter!

**Comment by Gr1N** [comment]
New error format is implemented, so I'm going to close this issue. All other improvements better to do in separate issues or pull requests.
